### ISSUE TITLE
Switch to opaque structs

### ISSUE THREAD

**Issue by michael-grunder** [valkey-io/libvalkey#39]
Right now, we declare `valkeyContext`, `valkeyAsyncContext`, `valkeyClusterContext`, and `valkeyAsyncClusterContext` in our public headers, exposing their memory layout.

This makes it a lot harder to make changes to them without breaking the ABI contract and therefore having to bump the SONAME.

We could switch to simply defining:

```c
typedef struct valkeyContext valkeyContext;
typedef struct valkeyClusterContext valkeyClusterContext;
```

in our public headers, and declare the actual struct internally in a private header.

The only real downside is that it might break code for users who are directly accessing struct members.  Personally, I can live with that :smile:.

@bjosv @zuiderkwast What does everybody think?

**Comment by zuiderkwast** [comment]
Björn is on vacation now. I say go for it!

Users have to migrate to this project anyway, at least the renaming parts. There are other changes too. We should collect all these changes for release notes and migration guide.

**Comment by zuiderkwast** [comment]
@bjosv says OK. If some user needs access to some field, we can add accessor functions for those later on.

**Comment by bjosv** [comment]
Yes, I like the idea. Lets add accessors when/if the time comes.
